sinks/httpsink.go

```diff
@@ -0,0 +1,161 @@
+(Hopefully in a single keep-alive http connection, which is go's default.)
+
+But with the payload of the messages being a serialized JSON object
+containing the kubernetes v1.Event.
+*/
+
+// HTTPSink wraps an HTTP endpoint that messages should be sent to.
+type HTTPSink struct {
+	SinkURL string
+
+	eventCh    channels.Channel
+	httpClient *pester.Client
+	bodyBuf    *bytes.Buffer
+}
+
+// NewHTTPSink constructs a new HTTPSink given a sink URL and buffer size
+func NewHTTPSink(sinkURL string, overflow bool, bufferSize int) *HTTPSink {
+	h := &HTTPSink{
+		SinkURL: sinkURL,
+	}
+
+	if overflow {
+		h.eventCh = channels.NewOverflowingChannel(channels.BufferCap(bufferSize))
+	} else {
+		h.eventCh = channels.NewNativeChannel(channels.BufferCap(bufferSize))
+	}
+
+	h.httpClient = pester.New()
+	h.httpClient.Backoff = pester.ExponentialJitterBackoff
+	h.httpClient.MaxRetries = 3
+	// Let the body buffer be 4096 bytes at the start. It will be grown if
+	// necessary.
+	h.bodyBuf = bytes.NewBuffer(make([]byte, 0, 4096))
+
+	return h
+}
+
+// UpdateEvents implements the EventSinkInterface. It really just writes the
+// event data to the event OverflowingChannel, which should never block.
+// Messages that are buffered beyond the bufferSize specified for this HTTPSink
+// are discarded.
+func (h *HTTPSink) UpdateEvents(eNew *v1.Event, eOld *v1.Event) {
+	h.eventCh.In() <- NewEventData(eNew, eOld)
+}
+
+// Run sits in a loop, waiting for data to come in through h.eventCh,
+// and forwarding them to the HTTP sink. If multiple events have happened
+// between loop iterations, it puts all of them in one request instead of
+// making a single request per event.
+func (h *HTTPSink) Run(stopCh <-chan bool) {
+loop:
+	for {
+		select {
+		case e := <-h.eventCh.Out():
+			var evt EventData
+			var ok bool
+			if evt, ok = e.(EventData); !ok {
+				glog.Warningf("Invalid type sent through event channel: %T", e)
+				continue loop
+			}
+
+			// Start with just this event...
+			arr := []EventData{evt}
+
+			// Consume all buffered events into an array, in case more have been written
+			// since we last forwarded them
+			numEvents := h.eventCh.Len()
+			for i := 0; i < numEvents; i++ {
+				e := <-h.eventCh.Out()
+				if evt, ok = e.(EventData); ok {
+					arr = append(arr, evt)
+				} else {
+					glog.Warningf("Invalid type sent through event channel: %T", e)
+				}
+			}
+
+			h.drainEvents(arr)
+		case <-stopCh:
+			break loop
+		}
+	}
+}
+
+// drainEvents takes an array of event data and sends it to the receiving HTTP
+// server. This function is *NOT* re-entrant: it re-uses the same body buffer
+// for each call, truncating it each time to avoid extra memory allocations.
+func (h *HTTPSink) drainEvents(events []EventData) {
+	// Reuse the body buffer for each request
+	h.bodyBuf.Truncate(0)
+
+	var written int64
+	for _, evt := range events {
+		w, err := evt.WriteRFC5424(h.bodyBuf)
+		written += w
+		if err != nil {
```

Comment:
Looking at the source, it happens if either the io.Writer returns an error (which bytes.Buffer is guaranteed never to do, it panics if it can't allocate enough), or if the message can't be marshalled into a byte string (which is unrecoverable, and really only happens if app names/etc are too long.  See https://github.com/crewjam/rfc5424/blob/master/marshal.go#L90)